File: typo-tester-mariadb-java/src/java/testdb/MariaDbTypeTest.java
package testdb;

import org.junit.Test;
import testdb.mariatest.MariatestId;
import testdb.mariatest.MariatestRepoImpl;
import testdb.mariatest.MariatestRow;
import testdb.mariatestnull.MariatestnullRepoImpl;
import testdb.mariatestnull.MariatestnullRow;
import typo.data.maria.Inet4;
import typo.data.maria.Inet6;
import typo.data.maria.MariaSet;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.Year;
import java.util.List;
import java.util.Optional;

import static testdb.JsonEquals.assertJsonEquals;
import static org.junit.Assert.*;

/**
 * Tests for all MariaDB data types - equivalent to PostgreSQL ArrayTest for MariaDB.
 */
public class MariaDbTypeTest {
    private final MariatestRepoImpl mariatestRepo = new MariatestRepoImpl();
    private final MariatestnullRepoImpl mariatestnullRepo = new MariatestnullRepoImpl();

    static MariatestRow mariaTestRow() {
        return new MariatestRow(
            // Integer types
            (byte) 127,                                      // tinyint_col
            (short) 32767,                                   // smallint_col
            8388607,                                         // mediumint_col
            new MariatestId(42),                             // int_col (PK)
            9223372036854775807L,                            // bigint_col
            // Unsigned integer types
            (short) 255,                                     // tinyint_u_col
            65535,                                           // smallint_u_col
            16777215,                                        // mediumint_u_col
            4294967295L,                                     // int_u_col
            new BigInteger("18446744073709551615"),          // bigint_u_col
            // Decimal types
            new BigDecimal("12345.6789"),                    // decimal_col
            new BigDecimal("98765.4321"),                    // numeric_col
            3.14f,                                           // float_col
            2.718281828459045,                               // double_col
            // Boolean
            true,                                            // bool_col
            // Bit types
            new byte[]{(byte) 0xF0},                          // bit_col (BIT(8))
            new byte[]{(byte) 1},                             // bit1_col (BIT(1))
            // String types
            "fixed     ",                                    // char_col (CHAR(10))
            "variable",                                      // varchar_col
            "tiny text",                                     // tinytext_col
            "text content",                                  // text_col
            "medium text",                                   // mediumtext_col
            "long text",                                     // longtext_col
            // Binary types
            new byte[]{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A}, // binary_col
            new byte[]{(byte)0xDE, (byte)0xAD, (byte)0xBE, (byte)0xEF}, // varbinary_col
            new byte[]{0x01},                                // tinyblob_col
            new byte[]{0x02, 0x03},                          // blob_col
            new byte[]{0x04, 0x05, 0x06},                    // mediumblob_col
            new byte[]{0x07, 0x08, 0x09, 0x0A},              // longblob_col
            // Date/time types
            LocalDate.of(2024, 12, 2),                       // date_col
            LocalTime.of(12, 30, 45),                        // time_col
            LocalTime.of(12, 30, 45, 123456000),             // time_fsp_col
            LocalDateTime.of(2024, 12, 2, 14, 30, 45),       // datetime_col
            LocalDateTime.of(2024, 12, 2, 14, 30, 45, 123456000), // datetime_fsp_col
            LocalDateTime.of(2024, 12, 2, 10, 0, 0),         // timestamp_col (default)
            LocalDateTime.of(2024, 12, 2, 10, 0, 0, 123456000), // timestamp_fsp_col (default)
            Year.of(2024),                                   // year_col
            // Enum and set
            "a",                                             // enum_col (ENUM values: a, b, c, d)
            MariaSet.of("x", "y"),                           // set_col (SET values: x, y, z)
            // JSON
            "{\"key\": \"value\"}",                          // json_col
            // Network types
            new Inet4("192.168.1.1"),                        // inet4_col
            new Inet6("::1")                                 // inet6_col
        );
    }

    static MariatestnullRow mariatestnullRowEmpty() {
        return new MariatestnullRow(
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
            Optional.empty(), Optional.empty()
        );
    }

    static MariatestnullRow mariatestnullRowWithValues() {
        return new MariatestnullRow(
            // Integer types
            Optional.of((byte) 127),                          // tinyint_col
            Optional.of((short) 32767),                       // smallint_col
            Optional.of(8388607),                             // mediumint_col
            Optional.of(42),                                  // int_col
            Optional.of(9223372036854775807L),                // bigint_col
            // Unsigned integer types
            Optional.of((short) 255),                         // tinyint_u_col
            Optional.of(65535),                               // smallint_u_col
            Optional.of(16777215),                            // mediumint_u_col
            Optional.of(4294967295L),                         // int_u_col
            Optional.of(new BigInteger("18446744073709551615")), // bigint_u_col
            // Decimal types
            Optional.of(new BigDecimal("12345.6789")),        // decimal_col
            Optional.of(new BigDecimal("98765.4321")),        // numeric_col
            Optional.of(3.14f),                               // float_col
            Optional.of(2.718281828459045),                   // double_col
            // Boolean
            Optional.of(true),                                // bool_col
            // Bit types
            Optional.of(new byte[]{(byte) 0xF0}),             // bit_col (BIT(8))
            Optional.of(new byte[]{(byte) 1}),                // bit1_col
            // String types
            Optional.of("fixed     "),                        // char_col
            Optional.of("variable"),                          // varchar_col
            Optional.of("tiny text"),                         // tinytext_col
            Optional.of("text content"),                      // text_col
            Optional.of("medium text"),                       // mediumtext_col
            Optional.of("long text"),                         // longtext_col
            // Binary types
            Optional.of(new byte[]{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A}), // binary_col
            Optional.of(new byte[]{(byte)0xDE, (byte)0xAD, (byte)0xBE, (byte)0xEF}), // varbinary_col
            Optional.of(new byte[]{0x01}),                    // tinyblob_col
            Optional.of(new byte[]{0x02, 0x03}),              // blob_col
            Optional.of(new byte[]{0x04, 0x05, 0x06}),        // mediumblob_col
            Optional.of(new byte[]{0x07, 0x08, 0x09, 0x0A}),  // longblob_col
            // Date/time types
            Optional.of(LocalDate.of(2024, 12, 2)),           // date_col
            Optional.of(LocalTime.of(12, 30, 45)),            // time_col
            Optional.of(LocalTime.of(12, 30, 45, 123456000)), // time_fsp_col
            Optional.of(LocalDateTime.of(2024, 12, 2, 14, 30, 45)), // datetime_col
            Optional.of(LocalDateTime.of(2024, 12, 2, 14, 30, 45, 123456000)), // datetime_fsp_col
            Optional.of(LocalDateTime.of(2024, 12, 2, 10, 0, 0)), // timestamp_col
            Optional.of(LocalDateTime.of(2024, 12, 2, 10, 0, 0, 123456000)), // timestamp_fsp_col
            Optional.of(Year.of(2024)),                       // year_col
            // Enum and set
            Optional.of("a"),                                 // enum_col (ENUM values: a, b, c, d)
            Optional.of(MariaSet.of("x", "y")),               // set_col (SET values: x, y, z)
            // JSON
            Optional.of("{\"key\": \"value\"}"),              // json_col
            // Network types
            Optional.of(new Inet4("192.168.1.1")),            // inet4_col
            Optional.of(new Inet6("::1"))                     // inet6_col
        );
    }

    @Test
    public void canInsertMariatestRows() {
        WithConnection.run(c -> {
            var before = mariaTestRow();
            var after = mariatestRepo.insert(before, c);
            // Compare key scalar fields - timestamps will differ due to defaults
            assertEquals(before.tinyintCol(), after.tinyintCol());
            assertEquals(before.smallintCol(), after.smallintCol());
            assertEquals(before.mediumintCol(), after.mediumintCol());
            assertEquals(before.intCol(), after.intCol());
            assertEquals(before.bigintCol(), after.bigintCol());
            assertEquals(before.boolCol(), after.boolCol());
            assertEquals(before.varcharCol(), after.varcharCol());
            assertEquals(before.textCol(), after.textCol());
            assertEquals(before.dateCol(), after.dateCol());
            assertEquals(before.timeCol(), after.timeCol());
            assertEquals(before.yearCol(), after.yearCol());
            assertEquals(before.enumCol(), after.enumCol());
            assertEquals(before.jsonCol(), after.jsonCol());
            assertEquals(before.inet4Col(), after.inet4Col());
            assertEquals(before.inet6Col(), after.inet6Col());
        });
    }

    @Test
    public void canInsertNullMariatestnullRows() {
        WithConnection.run(c -> {
            var before = mariatestnullRowEmpty();
            var after = mariatestnullRepo.insert(before, c);
            assertJsonEquals(before, after);
        });
    }

    @Test
    public void canInsertNonNullMariatestnullRows() {
        WithConnection.run(c -> {
            var before = mariatestnullRowWithValues();
            var after = mariatestnullRepo.insert(before, c);
            // Verify key fields match
            assertEquals(before.tinyintCol(), after.tinyintCol());
            assertEquals(before.smallintCol(), after.smallintCol());
            assertEquals(before.intCol(), after.intCol());
            assertEquals(before.bigintCol(), after.bigintCol());
            assertEquals(before.boolCol(), after.boolCol());
            assertEquals(before.varcharCol(), after.varcharCol());
            assertEquals(before.dateCol(), after.dateCol());
            assertEquals(before.yearCol(), after.yearCol());
            assertEquals(before.enumCol(), after.enumCol());
            assertEquals(before.inet4Col(), after.inet4Col());
            assertEquals(before.inet6Col(), after.inet6Col());
        });
    }

    @Test
    public void canSelectAllMariatestRows() {
        WithConnection.run(c -> {
            var row1 = mariaTestRow();
            var row2 = mariaTestRow().withIntCol(new MariatestId(43));
            mariatestRepo.insert(row1, c);
            mariatestRepo.insert(row2, c);
            var all = mariatestRepo.selectAll(c);
            assertEquals(2, all.size());
        });
    }

    @Test
    public void canSelectByIdMariatestRows() {
        WithConnection.run(c -> {
            var inserted = mariatestRepo.insert(mariaTestRow(), c);
            var selected = mariatestRepo.selectById(inserted.intCol(), c);
            assertTrue(selected.isPresent());
            assertEquals(inserted.intCol(), selected.get().intCol());
        });
    }

    @Test
    public void canSelectByIdsMariatestRows() {
        WithConnection.run(c -> {
            var row1 = mariatestRepo.insert(mariaTestRow(), c);
            var row2 = mariatestRepo.insert(mariaTestRow().withIntCol(new MariatestId(43)), c);
            var ids = new MariatestId[]{row1.intCol(), row2.intCol()};
            var selected = mariatestRepo.selectByIds(ids, c);
            assertEquals(2, selected.size());
        });
    }

    @Test
    public void canUpdateMariatestRows() {
        WithConnection.run(c -> {
            var inserted = mariatestRepo.insert(mariaTestRow(), c);
            var updated = inserted.withVarcharCol("updated");
            mariatestRepo.update(updated, c);
            var selected = mariatestRepo.selectById(inserted.intCol(), c);
            assertTrue(selected.isPresent());
            assertEquals("updated", selected.get().varcharCol());
        });
    }

    @Test
    public void canDeleteMariatestRows() {
        WithConnection.run(c -> {
            var inserted = mariatestRepo.insert(mariaTestRow(), c);
            var deleted = mariatestRepo.deleteById(inserted.intCol(), c);
            assertTrue(deleted);
            var selected = mariatestRepo.selectById(inserted.intCol(), c);
            assertFalse(selected.isPresent());
        });
    }

    @Test
    public void canDeleteByIdsMariatestRows() {
        WithConnection.run(c -> {
            var row1 = mariatestRepo.insert(mariaTestRow(), c);
            var row2 = mariatestRepo.insert(mariaTestRow().withIntCol(new MariatestId(43)), c);
            var ids = new MariatestId[]{row1.intCol(), row2.intCol()};
            var count = mariatestRepo.deleteByIds(ids, c);
            assertEquals(Integer.valueOf(2), count);
            var all = mariatestRepo.selectAll(c);
            assertEquals(0, all.size());
        });
    }

    @Test
    public void canUpsertMariatestRows() {
        WithConnection.run(c -> {
            var row = mariaTestRow();
            var inserted = mariatestRepo.upsert(row, c);
            assertEquals(row.intCol(), inserted.intCol());

            var updated = inserted.withVarcharCol("upserted");
            var upserted = mariatestRepo.upsert(updated, c);
            assertEquals("upserted", upserted.varcharCol());
        });
    }

    @Test
    public void canQueryMariatestnullWithDSL() {
        WithConnection.run(c -> {
            var row = mariatestnullRepo.insert(mariatestnullRowWithValues(), c);

            // Test DSL select.where() for various types
            assertEquals(row.boolCol(), mariatestnullRepo.select().where(p -> p.boolCol().isEqual(row.boolCol().orElse(null))).toList(c).get(0).boolCol());
            assertEquals(row.varcharCol(), mariatestnullRepo.select().where(p -> p.varcharCol().isEqual(row.varcharCol().orElse(null))).toList(c).get(0).varcharCol());
            assertEquals(row.intCol(), mariatestnullRepo.select().where(p -> p.intCol().isEqual(row.intCol().orElse(null))).toList(c).get(0).intCol());
            assertEquals(row.bigintCol(), mariatestnullRepo.select().where(p -> p.bigintCol().isEqual(row.bigintCol().orElse(null))).toList(c).get(0).bigintCol());
            assertEquals(row.dateCol(), mariatestnullRepo.select().where(p -> p.dateCol().isEqual(row.dateCol().orElse(null))).toList(c).get(0).dateCol());
            assertEquals(row.yearCol(), mariatestnullRepo.select().where(p -> p.yearCol().isEqual(row.yearCol().orElse(null))).toList(c).get(0).yearCol());
            assertEquals(row.enumCol(), mariatestnullRepo.select().where(p -> p.enumCol().isEqual(row.enumCol().orElse(null))).toList(c).get(0).enumCol());
        });
    }

    @Test
    public void canQueryMariatestWithDSL() {
        WithConnection.run(c -> {
            var row = mariatestRepo.insert(mariaTestRow(), c);

            // Test DSL update.setValue() for various types
            mariatestRepo.update().setValue(p -> p.boolCol(), row.boolCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.varcharCol(), row.varcharCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.textCol(), row.textCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.bigintCol(), row.bigintCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.dateCol(), row.dateCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.timeCol(), row.timeCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.yearCol(), row.yearCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.enumCol(), row.enumCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
            mariatestRepo.update().setValue(p -> p.jsonCol(), row.jsonCol()).where(p -> p.intCol().isEqual(row.intCol())).execute(c);
        });
    }
}

File: typo-tester-mariadb-kotlin/src/kotlin/testdb/MariaDbTypeTest.kt
package testdb

import org.junit.Assert.*
import org.junit.Test
import testdb.JsonEquals.assertJsonEquals
import testdb.mariatest.MariatestId
import testdb.mariatest.MariatestRepoImpl
import testdb.mariatest.MariatestRow
import testdb.mariatestnull.MariatestnullRepoImpl
import testdb.mariatestnull.MariatestnullRow
import typo.data.maria.Inet4
import typo.data.maria.Inet6
import typo.data.maria.MariaSet
import java.math.BigDecimal
import java.math.BigInteger
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.Year

/**
 * Tests for all MariaDB data types - equivalent to PostgreSQL ArrayTest for MariaDB.
 */
class MariaDbTypeTest {
    private val mariatestRepo = MariatestRepoImpl()
    private val mariatestnullRepo = MariatestnullRepoImpl()

    companion object {
        fun mariaTestRow(): MariatestRow = MariatestRow(
            // Integer types
            tinyintCol = 127.toByte(),
            smallintCol = 32767.toShort(),
            mediumintCol = 8388607,
            intCol = MariatestId(42),
            bigintCol = 9223372036854775807L,
            // Unsigned integer types
            tinyintUCol = 255.toShort(),
            smallintUCol = 65535,
            mediumintUCol = 16777215,
            intUCol = 4294967295L,
            bigintUCol = BigInteger("18446744073709551615"),
            // Decimal types
            decimalCol = BigDecimal("12345.6789"),
            numericCol = BigDecimal("98765.4321"),
            floatCol = 3.14f,
            doubleCol = 2.718281828459045,
            // Boolean
            boolCol = true,
            // Bit types
            bitCol = byteArrayOf(0xF0.toByte()),
            bit1Col = byteArrayOf(1.toByte()),
            // String types
            charCol = "fixed     ",
            varcharCol = "variable",
            tinytextCol = "tiny text",
            textCol = "text content",
            mediumtextCol = "medium text",
            longtextCol = "long text",
            // Binary types
            binaryCol = byteArrayOf(0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A),
            varbinaryCol = byteArrayOf(0xDE.toByte(), 0xAD.toByte(), 0xBE.toByte(), 0xEF.toByte()),
            tinyblobCol = byteArrayOf(0x01),
            blobCol = byteArrayOf(0x02, 0x03),
            mediumblobCol = byteArrayOf(0x04, 0x05, 0x06),
            longblobCol = byteArrayOf(0x07, 0x08, 0x09, 0x0A),
            // Date/time types
            dateCol = LocalDate.of(2024, 12, 2),
            timeCol = LocalTime.of(12, 30, 45),
            timeFspCol = LocalTime.of(12, 30, 45, 123456000),
            datetimeCol = LocalDateTime.of(2024, 12, 2, 14, 30, 45),
            datetimeFspCol = LocalDateTime.of(2024, 12, 2, 14, 30, 45, 123456000),
            timestampCol = LocalDateTime.of(2024, 12, 2, 10, 0, 0),
            timestampFspCol = LocalDateTime.of(2024, 12, 2, 10, 0, 0, 123456000),
            yearCol = Year.of(2024),
            // Enum and set
            enumCol = "a",
            setCol = MariaSet.of("x", "y"),
            // JSON
            jsonCol = "{\"key\": \"value\"}",
            // Network types
            inet4Col = Inet4("192.168.1.1"),
            inet6Col = Inet6("::1")
        )

        fun mariatestnullRowEmpty(): MariatestnullRow = MariatestnullRow(
            tinyintCol = null,
            smallintCol = null,
            mediumintCol = null,
            intCol = null,
            bigintCol = null,
            tinyintUCol = null,
            smallintUCol = null,
            mediumintUCol = null,
            intUCol = null,
            bigintUCol = null,
            decimalCol = null,
            numericCol = null,
            floatCol = null,
            doubleCol = null,
            boolCol = null,
            bitCol = null,
            bit1Col = null,
            charCol = null,
            varcharCol = null,
            tinytextCol = null,
            textCol = null,
            mediumtextCol = null,
            longtextCol = null,
            binaryCol = null,
            varbinaryCol = null,
            tinyblobCol = null,
            blobCol = null,
            mediumblobCol = null,
            longblobCol = null,
            dateCol = null,
            timeCol = null,
            timeFspCol = null,
            datetimeCol = null,
            datetimeFspCol = null,
            timestampCol = null,
            timestampFspCol = null,
            yearCol = null,
            enumCol = null,
            setCol = null,
            jsonCol = null,
            inet4Col = null,
            inet6Col = null
        )

        fun mariatestnullRowWithValues(): MariatestnullRow = MariatestnullRow(
            // Integer types
            tinyintCol = 127.toByte(),
            smallintCol = 32767.toShort(),
            mediumintCol = 8388607,
            intCol = 42,
            bigintCol = 9223372036854775807L,
            // Unsigned integer types
            tinyintUCol = 255.toShort(),
            smallintUCol = 65535,
            mediumintUCol = 16777215,
            intUCol = 4294967295L,
            bigintUCol = BigInteger("18446744073709551615"),
            // Decimal types
            decimalCol = BigDecimal("12345.6789"),
            numericCol = BigDecimal("98765.4321"),
            floatCol = 3.14f,
            doubleCol = 2.718281828459045,
            // Boolean
            boolCol = true,
            // Bit types
            bitCol = byteArrayOf(0xF0.toByte()),
            bit1Col = byteArrayOf(1.toByte()),
            // String types
            charCol = "fixed     ",
            varcharCol = "variable",
            tinytextCol = "tiny text",
            textCol = "text content",
            mediumtextCol = "medium text",
            longtextCol = "long text",
            // Binary types
            binaryCol = byteArrayOf(0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A),
            varbinaryCol = byteArrayOf(0xDE.toByte(), 0xAD.toByte(), 0xBE.toByte(), 0xEF.toByte()),
            tinyblobCol = byteArrayOf(0x01),
            blobCol = byteArrayOf(0x02, 0x03),
            mediumblobCol = byteArrayOf(0x04, 0x05, 0x06),
            longblobCol = byteArrayOf(0x07, 0x08, 0x09, 0x0A),
            // Date/time types
            dateCol = LocalDate.of(2024, 12, 2),
            timeCol = LocalTime.of(12, 30, 45),
            timeFspCol = LocalTime.of(12, 30, 45, 123456000),
            datetimeCol = LocalDateTime.of(2024, 12, 2, 14, 30, 45),
            datetimeFspCol = LocalDateTime.of(2024, 12, 2, 14, 30, 45, 123456000),
            timestampCol = LocalDateTime.of(2024, 12, 2, 10, 0, 0),
            timestampFspCol = LocalDateTime.of(2024, 12, 2, 10, 0, 0, 123456000),
            yearCol = Year.of(2024),
            // Enum and set
            enumCol = "a",
            setCol = MariaSet.of("x", "y"),
            // JSON
            jsonCol = "{\"key\": \"value\"}",
            // Network types
            inet4Col = Inet4("192.168.1.1"),
            inet6Col = Inet6("::1")
        )
    }

    @Test
    fun canInsertMariatestRows() {
        WithConnection.run { c ->
            val before = mariaTestRow()
            val after = mariatestRepo.insert(before, c)
            // Compare key scalar fields - timestamps will differ due to defaults
            assertEquals(before.tinyintCol, after.tinyintCol)
            assertEquals(before.smallintCol, after.smallintCol)
            assertEquals(before.mediumintCol, after.mediumintCol)
            assertEquals(before.intCol, after.intCol)
            assertEquals(before.bigintCol, after.bigintCol)
            assertEquals(before.boolCol, after.boolCol)
            assertEquals(before.varcharCol, after.varcharCol)
            assertEquals(before.textCol, after.textCol)
            assertEquals(before.dateCol, after.dateCol)
            assertEquals(before.timeCol, after.timeCol)
            assertEquals(before.yearCol, after.yearCol)
            assertEquals(before.enumCol, after.enumCol)
            assertEquals(before.jsonCol, after.jsonCol)
            assertEquals(before.inet4Col, after.inet4Col)
            assertEquals(before.inet6Col, after.inet6Col)
        }
    }

    @Test
    fun canInsertNullMariatestnullRows() {
        WithConnection.run { c ->
            val before = mariatestnullRowEmpty()
            val after = mariatestnullRepo.insert(before, c)
            assertJsonEquals(before, after)
        }
    }

    @Test
    fun canInsertNonNullMariatestnullRows() {
        WithConnection.run { c ->
            val before = mariatestnullRowWithValues()
            val after = mariatestnullRepo.insert(before, c)
            // Verify key fields match
            assertEquals(before.tinyintCol, after.tinyintCol)
            assertEquals(before.smallintCol, after.smallintCol)
            assertEquals(before.intCol, after.intCol)
            assertEquals(before.bigintCol, after.bigintCol)
            assertEquals(before.boolCol, after.boolCol)
            assertEquals(before.varcharCol, after.varcharCol)
            assertEquals(before.dateCol, after.dateCol)
            assertEquals(before.yearCol, after.yearCol)
            assertEquals(before.enumCol, after.enumCol)
            assertEquals(before.inet4Col, after.inet4Col)
            assertEquals(before.inet6Col, after.inet6Col)
        }
    }

    @Test
    fun canSelectAllMariatestRows() {
        WithConnection.run { c ->
            val row1 = mariaTestRow()
            val row2 = mariaTestRow().copy(intCol = MariatestId(43))
            mariatestRepo.insert(row1, c)
            mariatestRepo.insert(row2, c)
            val all = mariatestRepo.selectAll(c)
            assertEquals(2, all.size)
        }
    }

    @Test
    fun canSelectByIdMariatestRows() {
        WithConnection.run { c ->
            val inserted = mariatestRepo.insert(mariaTestRow(), c)
            val selected = mariatestRepo.selectById(inserted.intCol, c)
            assertNotNull(selected)
            assertEquals(inserted.intCol, selected!!.intCol)
        }
    }

    @Test
    fun canSelectByIdsMariatestRows() {
        WithConnection.run { c ->
            val row1 = mariatestRepo.insert(mariaTestRow(), c)
            val row2 = mariatestRepo.insert(mariaTestRow().copy(intCol = MariatestId(43)), c)
            val ids = arrayOf(row1.intCol, row2.intCol)
            val selected = mariatestRepo.selectByIds(ids, c)
            assertEquals(2, selected.size)
        }
    }

    @Test
    fun canUpdateMariatestRows() {
        WithConnection.run { c ->
            val inserted = mariatestRepo.insert(mariaTestRow(), c)
            val updated = inserted.copy(varcharCol = "updated")
            mariatestRepo.update(updated, c)
            val selected = mariatestRepo.selectById(inserted.intCol, c)
            assertNotNull(selected)
            assertEquals("updated", selected!!.varcharCol)
        }
    }

    @Test
    fun canDeleteMariatestRows() {
        WithConnection.run { c ->
            val inserted = mariatestRepo.insert(mariaTestRow(), c)
            val deleted = mariatestRepo.deleteById(inserted.intCol, c)
            assertTrue(deleted)
            val selected = mariatestRepo.selectById(inserted.intCol, c)
            assertNull(selected)
        }
    }

    @Test
    fun canDeleteByIdsMariatestRows() {
        WithConnection.run { c ->
            val row1 = mariatestRepo.insert(mariaTestRow(), c)
            val row2 = mariatestRepo.insert(mariaTestRow().copy(intCol = MariatestId(43)), c)
            val ids = arrayOf(row1.intCol, row2.intCol)
            val count = mariatestRepo.deleteByIds(ids, c)
            assertEquals(2, count)
            val all = mariatestRepo.selectAll(c)
            assertEquals(0, all.size)
        }
    }

    @Test
    fun canUpsertMariatestRows() {
        WithConnection.run { c ->
            val row = mariaTestRow()
            val inserted = mariatestRepo.upsert(row, c)
            assertEquals(row.intCol, inserted.intCol)

            val updated = inserted.copy(varcharCol = "upserted")
            val upserted = mariatestRepo.upsert(updated, c)
            assertEquals("upserted", upserted.varcharCol)
        }
    }

    @Test
    fun canQueryMariatestnullWithDSL() {
        WithConnection.run { c ->
            val row = mariatestnullRepo.insert(mariatestnullRowWithValues(), c)

            // Test DSL select.where() for various types (use !! since we know the test data has non-null values)
            assertEquals(row.boolCol, mariatestnullRepo.select().where { p -> p.boolCol().isEqual(row.boolCol!!) }.toList(c)[0].boolCol)
            assertEquals(row.varcharCol, mariatestnullRepo.select().where { p -> p.varcharCol().isEqual(row.varcharCol!!) }.toList(c)[0].varcharCol)
            assertEquals(row.intCol, mariatestnullRepo.select().where { p -> p.intCol().isEqual(row.intCol!!) }.toList(c)[0].intCol)
            assertEquals(row.bigintCol, mariatestnullRepo.select().where { p -> p.bigintCol().isEqual(row.bigintCol!!) }.toList(c)[0].bigintCol)
            assertEquals(row.dateCol, mariatestnullRepo.select().where { p -> p.dateCol().isEqual(row.dateCol!!) }.toList(c)[0].dateCol)
            assertEquals(row.yearCol, mariatestnullRepo.select().where { p -> p.yearCol().isEqual(row.yearCol!!) }.toList(c)[0].yearCol)
            assertEquals(row.enumCol, mariatestnullRepo.select().where { p -> p.enumCol().isEqual(row.enumCol!!) }.toList(c)[0].enumCol)
        }
    }

    @Test
    fun canQueryMariatestWithDSL() {
        WithConnection.run { c ->
            val row = mariatestRepo.insert(mariaTestRow(), c)

            // Test DSL update.setValue() for various types
            mariatestRepo.update().setValue({ p -> p.boolCol() }, row.boolCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.varcharCol() }, row.varcharCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.textCol() }, row.textCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.bigintCol() }, row.bigintCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.dateCol() }, row.dateCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.timeCol() }, row.timeCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.yearCol() }, row.yearCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.enumCol() }, row.enumCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
            mariatestRepo.update().setValue({ p -> p.jsonCol() }, row.jsonCol).where { p -> p.intCol().isEqual(row.intCol) }.execute(c)
        }
    }
}
